### ISSUE TITLE
Request for comments: Support for shell autocompletion

### ISSUE THREAD

**Issue by ncreuschling** [dry-rb/dry-cli#123]
I would like to get your feedback whether you would accept a PR for shell autocompletion.

I am in the process of migrating a RubyGems from [gli](http://davetron5000.github.io/gli/) to dry-cli. gli generates output for bash and zsh autocompletion e.g. `rubygem-command help -c`.

I am pretty sure `hanami-cli` would profit from such a feature as well.

Should I take a shot at this or is this something you are not interested?

**Comment by timriley** [comment]
Yep, I'd definitely be happy to see something like this be added!

**Comment by ncreuschling** [comment]
Note to self: This is somewhat tricky. Commands are not readily available as a list. CommandRegistry is tree-ish.

**Comment by kigster** [comment]
I built this as a gem, and I would rather see it land here than live outside.

**dry-cli-autocomplete** ([GitHub](https://github.com/kigster/dry-cli-autocomplete), [RubyGems](https://rubygems.org/gems/dry-cli-autocomplete)) generates a static completion script from a registry. A host registers one command, `mycli completion bash` or `mycli completion zsh` prints a script, and the user sources it from a shell profile.

```ruby
require "dry/cli/autocomplete/command"

module MyCLI
  extend Dry::CLI::Registry

  register "completion", Dry::CLI::Autocomplete::Command[MyCLI]
end
```

### Why static, and not a callback

@gustavothecoder's #141 routes each TAB press back into the CLI, the way Cobra and clap do. That is the right call for a Go binary that starts in 10ms, and I went in expecting to do the same. The timings talked me out of it, profiling a 27-command app of mine:

| | |
|---|---:|
| Bare `ruby -e ''` | 100ms |
| `require "dry/cli"` | 160ms |
| `require` a heavy host | 520ms |
| Walking the registry and building the whole completion spec | **0.067ms** |

The Ruby process is the entire cost. A user holding TAB down gets half a second of dead air per keystroke and concludes the CLI is broken. Since the work itself is 0.067ms, generating everything once and writing it to a file costs nothing anyone can perceive, and TAB then spawns no process at all.

So I think these are different features rather than competing ones, and #141's approach is what you want if completions ever need to depend on runtime state. I would not want my thing to be the reason that PR keeps waiting.

#### The four things it does not handle

 [listed in my README](https://github.com/kigster/dry-cli-autocomplete#the-problem), are what I set out to fix.

### What the new gem covers

Beyond command and subcommand names: options and their aliases, values declared with `values:`, arguments that take a path (real `compgen -f` / `_files`), and it skips commands registered `hidden: true`. zsh gets a native `#compdef` script built on `_arguments` and `_describe`, carrying each option's `desc` as help text, rather than a `bashcompinit` shim.

Tests run the generated scripts through `bash -n` and `zsh -n`, and drive real completions by setting `COMP_WORDS` and reading `COMPREPLY` back, against three fixture registries. Runtime dependencies are dry-cli and dry-inflector.

### Three things I would like your read on

**The namespace.** I published as `dry-cli-autocomplete` under `Dry::CLI::Autocomplete`, which borrows your name without asking. If you would rather I not, I will rename it. That question is really why I am posting.

**Whether you want it here.** I am happy to port it into dry-cli as a PR, hand the gem to the org, or keep maintaining it outside with a clear note that it is unaffiliated. Any of those is fine by me.

**It reads private API.** Walking a registry means `Registry#get`, `CommandRegistry`, every `LookupResult` reader and every `Node` reader, all marked `@api private`. I avoided `instance_variable_get`, but a dry-cli release can still break the walk, and my fixture suite is the only thing that would catch it. That argues for the code living in this repo rather than guessing at it from outside.

Related: #39 was closed pointing at rngtng/dry-cli-completion. I read that gem first and it is where this started. The four things it does not handle, listed in my README, are what I set out to fix.

## AI Use

This was written in a pair programming session with Claude, where Claude especially helped with ZSH autocompletion. Everything wad reviwed by a human and mostly written by one too.

**Comment by timriley** [comment]
I'm still interested in having shell autocomplete support built into Dry CLI.

@kigster — I welcome your interest here, but based on your comments above, I want to point you to Hanakai's [AI contributions policy](https://hanakai.org/ai-policy). We want to work with people who put real attention into our ecosystem, including in how they write about it. I'm not interested in interacting with other people through the unadulterated outputs of their LLM prompts. If you'd like to have another go at this after reading that policy, please let me know!

Otherwise, we'll get to this feature when we can. We're currently preparing for a Dry CLI 2.0 release with a range of other changes. I would expect autocomplete support would come some time after that.

**Comment by kigster** [comment]
@timriley I wrote most of the Ruby code, got a bit of help with ZSH autocomplete because I am less familiar with it. The rest of the Ruby is trivial and did not need the AI at all, although these days I push PRs from Claude code so it can save me time writing commit descriptions. That's reason commits are shown as co-authored with Claude.

I am fine for this to exist as a standalone gem, i already integrated it into three of my dry-CLI based tools and it's a night and day difference working with them, especially in ZSH where it prints command and descriptions next to it. 

I am excited to hear about 2.0. 

My personal feature request is a custom top level banner for the entire tool, which shows up above the command list.

Another one is automatic spelling correction. If my command is `law` and I type `lwa` it would be nice if it just executed `law` with a warning message. 

I am currently building a web API service for internal tax law attestations. As part of that I was able to design, write and convert all IRS instructions to a completely custom language written in Rust in one evening. I am curious about the reasons behind this rule and for how long it will stand. Because I doubt zero AI policy will serve any project well long term. Maybe I am wrong. But that is just my opinion.

**Comment by timriley** [comment]
Hi @kigster, thanks for your reply!

I think you must have misunderstood our AI contributions policy. Nowhere there does it say "Zero AI".

Let me explicitly quote the parts that applied to your original comment (emphasis mine):

> - You must carefully read and review all LLM-generated code or text before asking maintainers to review it. **This includes PR descriptions, where we want to hear your personal voice rather than an unfiltered AI summary.**
> - You should be transparent and make note if your contribution contains substantial amounts of tool-generated content.

Your original comment was clearly LLM-generated, and did not feel like I was interacting with another human. I feel like I already know you much better simply from your reply above, which reads as if you wrote it yourself, not Claude. (This is the whole idea behind "here for humans" at the top of our policy — our whole point in OSS collaboration is to build connections with other people, and help build their direct personal knowledge of our tools).

In your comment above you also disclosed the level to which you used LLM-assistance. So you've pretty much satisfied the policy now. Thank you!

Thanks also for your thoughts about other improvements to Dry CLI!

**Comment by kigster** [comment]
@timriley You are right! I should have read it more carefully. It's a totally reasonable  policy and I have zero qualms with it, having read it 😂

Thanks for dry-rb project, it's my default go to set for most Ruby tools. While most of the jobs I worked at were Rails (starting in 2007) most of my open source work, infrastructure CLIs, and internal tools are rails-free. 

I am excited to try Hanami next as my API layer for tax agents to attest (sign off) on the chunks of law. It's not a very wide api so Hanami should fit the bill.
